llvm/test/CodeGen/NVPTX/fma-relu-contract.ll

```diff
@@ -0,0 +1,1422 @@
+; CHECK-SM70-NEXT:    mov.b32 {%rs1, %rs2}, %r3;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r4, %rs1;
+; CHECK-SM70-NEXT:    shl.b32 %r5, %r4, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f1, %r5;
+; CHECK-SM70-NEXT:    mov.b32 {%rs4, %rs5}, %r2;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r6, %rs4;
+; CHECK-SM70-NEXT:    shl.b32 %r7, %r6, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f2, %r7;
+; CHECK-SM70-NEXT:    mov.b32 {%rs7, %rs8}, %r1;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r8, %rs7;
+; CHECK-SM70-NEXT:    shl.b32 %r9, %r8, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f3, %r9;
+; CHECK-SM70-NEXT:    fma.rn.f32 %f4, %f3, %f2, %f1;
+; CHECK-SM70-NEXT:    mov.b32 %r10, %f4;
+; CHECK-SM70-NEXT:    bfe.u32 %r11, %r10, 16, 1;
+; CHECK-SM70-NEXT:    add.s32 %r12, %r11, %r10;
+; CHECK-SM70-NEXT:    add.s32 %r13, %r12, 32767;
+; CHECK-SM70-NEXT:    setp.nan.f32 %p1, %f4, %f4;
+; CHECK-SM70-NEXT:    or.b32 %r14, %r10, 4194304;
+; CHECK-SM70-NEXT:    selp.b32 %r15, %r14, %r13, %p1;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r16, %rs2;
+; CHECK-SM70-NEXT:    shl.b32 %r17, %r16, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f5, %r17;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r18, %rs5;
+; CHECK-SM70-NEXT:    shl.b32 %r19, %r18, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f6, %r19;
+; CHECK-SM70-NEXT:    cvt.u32.u16 %r20, %rs8;
+; CHECK-SM70-NEXT:    shl.b32 %r21, %r20, 16;
+; CHECK-SM70-NEXT:    mov.b32 %f7, %r21;
+; CHECK-SM70-NEXT:    fma.rn.f32 %f8, %f7, %f6, %f5;
+; CHECK-SM70-NEXT:    mov.b32 %r22, %f8;
+; CHECK-SM70-NEXT:    bfe.u32 %r23, %r22, 16, 1;
+; CHECK-SM70-NEXT:    add.s32 %r24, %r23, %r22;
+; CHECK-SM70-NEXT:    add.s32 %r25, %r24, 32767;
+; CHECK-SM70-NEXT:    setp.nan.f32 %p2, %f8, %f8;
+; CHECK-SM70-NEXT:    or.b32 %r26, %r22, 4194304;
+; CHECK-SM70-NEXT:    selp.b32 %r27, %r26, %r25, %p2;
+; CHECK-SM70-NEXT:    and.b32 %r28, %r27, -65536;
+; CHECK-SM70-NEXT:    mov.b32 %f9, %r28;
+; CHECK-SM70-NEXT:    max.f32 %f10, %f9, 0f00000000;
+; CHECK-SM70-NEXT:    mov.b32 %r29, %f10;
+; CHECK-SM70-NEXT:    bfe.u32 %r30, %r29, 16, 1;
+; CHECK-SM70-NEXT:    add.s32 %r31, %r30, %r29;
+; CHECK-SM70-NEXT:    add.s32 %r32, %r31, 32767;
+; CHECK-SM70-NEXT:    setp.nan.f32 %p3, %f10, %f10;
+; CHECK-SM70-NEXT:    or.b32 %r33, %r29, 4194304;
+; CHECK-SM70-NEXT:    selp.b32 %r34, %r33, %r32, %p3;
+; CHECK-SM70-NEXT:    { .reg .b16 tmp; mov.b32 {tmp, %rs13}, %r34; }
+; CHECK-SM70-NEXT:    and.b32 %r35, %r15, -65536;
+; CHECK-SM70-NEXT:    mov.b32 %f11, %r35;
+; CHECK-SM70-NEXT:    max.f32 %f12, %f11, 0f00000000;
+; CHECK-SM70-NEXT:    mov.b32 %r36, %f12;
+; CHECK-SM70-NEXT:    bfe.u32 %r37, %r36, 16, 1;
+; CHECK-SM70-NEXT:    add.s32 %r38, %r37, %r36;
+; CHECK-SM70-NEXT:    add.s32 %r39, %r38, 32767;
+; CHECK-SM70-NEXT:    setp.nan.f32 %p4, %f12, %f12;
+; CHECK-SM70-NEXT:    or.b32 %r40, %r36, 4194304;
+; CHECK-SM70-NEXT:    selp.b32 %r41, %r40, %r39, %p4;
+; CHECK-SM70-NEXT:    { .reg .b16 tmp; mov.b32 {tmp, %rs15}, %r41; }
+; CHECK-SM70-NEXT:    mov.b32 %r42, {%rs15, %rs13};
```

Comment:
Another future optimization opportunity -- we could lower those "truncate-to-i16, then recombine into i32" as a single `prmt`